### PR TITLE
[PERF] reduce peak memory in _build_indexer_reorder_contents

### LINKED ISSUE #1655: [PERF] Build pivot_longer reorder indexer with one NumPy allocation
## Summary

`_build_indexer_reorder_contents` currently constructs `np.arange(length * reps)`, reshapes it, and ravels in Fortran order. Because the final ravel requires a reordered copy, the helper temporarily holds roughly twice the output allocation.

A mathematically equivalent NumPy formulation can construct the row-by-column offsets directly:

```python
rows = np.arange(length)[:, None]
columns = np.arange(reps) * length
indexer = np.add(rows, columns).ravel()
```

This keeps the implementation in NumPy, preserves the current positional contract, and substantially lowers peak temporary memory.

## Measurements

Same-process medians on pandas 3.0.5 / NumPy 2.3.5:

| Shape | Current | One-allocation formulation | Current peak | New peak |
| --- | ---: | ---: | ---: | ---: |
| `length=100_000`, `reps=20` (2M output) | 1.49 ms | 1.49 ms | 30.52 MiB | 16.15 MiB |
| `length=1_000`, `reps=1_000` (1M output) | 1.09 ms | 0.63 ms | 15.26 MiB | 7.77 MiB |
| `length=1_000_000`, `reps=8` (8M output) | 10.53 ms | 8.11 ms | 122.07 MiB | 68.79 MiB |

The generated arrays were element-for-element equivalent.

## Scope

- Replace only `_build_indexer_reorder_contents`.
- Add focused helper equivalence tests for zero/small sizes and both row-heavy and column-heavy shapes.
- Confirm public `pivot_longer` behavior for extension arrays with `sort_by_appearance=True`.
- Measure end-to-end runtime and peak memory on many-row and many-column workloads.
- Preserve `np.intp`-compatible indexing and overflow/error behavior.

## Acceptance criteria

- Exact positional equivalence with the current helper.
- No ordering or dtype regression across nullable integer, string, categorical, datetime, and timezone-aware extension arrays.
- Peak helper allocation materially decreases for large outputs.
- No meaningful runtime regression for small or representative workloads.
- Complete pivot tests pass under supported pandas environments.

## Dependencies

Land after #1636 / PR #1643 so its allocation changes remain independently reviewable. This NumPy baseline should precede any Rust experiment in pyjanitor-devs/janitor-rs#18.

### PR DESCRIPTION
**Pull request by Anupam2400**
## PR Description

- Replace `_build_indexer_reorder_contents` double allocation with
  single NumPy allocation using direct row+column offset construction
- Old: `np.arange(length*reps).reshape().ravel('F')` creates temp copy
- New: `np.add(rows, columns).ravel()` constructs output directly

**Benchmark (length=1M, reps=8 → 8M output):**

| Metric | Current | New | Change |
|---|---|---|---|
| Peak memory | 122.07 MiB | 68.79 MiB | **43.6% less** |
| Runtime (median of 15 runs, `timeit`, GC disabled) | 16.41 ms | 16.98 ms | ~3.5% slower, within measurement noise |

This PR is a memory-efficiency optimization. An earlier version of this
description claimed a 1.52x runtime speedup — that number came from a
single untimed cold measurement (`time.time()` around one call) and did
not hold up under repeated, controlled benchmarking. Runtime is
essentially unchanged between the two implementations, possibly
marginally slower, likely since the new path builds the result via
broadcasting rather than a single `arange` + `reshape`. The memory
savings, however, are real, reproducible across multiple scales
(10M×4, 1M×16, 100×100k), and the actual motivation for this change.

**Tests added:**

- Zero length edge case
- Zero reps edge case (defensive coverage; not reachable via the
  public API today, but locked in in case that changes)
- Single rep / single length edge cases
- Row-heavy (1000×2) and column-heavy (2×1000) equivalence
- Peak memory assertion (< 70 MiB)
- End-to-end through `pivot_longer(..., sort_by_appearance=True)`
  with nullable Int64, string, categorical, and datetime columns,
  verifying both values and dtypes survive the reorder

All 101 `test_pivot_longer` tests pass ✅

**This PR resolves #1655.**

## PR Checklist

- [x] PR in from a fork off your branch.
- [x] Add yourself to `AUTHORS.md`.
- [x] Add a line to `CHANGELOG.md` under the latest version header.

## Relevant Reviewers

@samukweku

### CODE REVIEW

**Comment by samukweku** [comment]
`End-to-end with nullable int, string, categorical, datetime arrays`

@Anupam2400 kindly add tests that directly cover this

**Comment by samukweku** [comment]
@Anupam2400 how was the 1.52x computed?

**Comment by Anupam2400** [comment]
@samukweku

The 1.52x was from a single cold `time.time()` call and doesn't hold
up under scrutiny. Re-ran with `timeit.repeat()` (median of 15 runs ×
200 calls, GC disabled) on both implementations: old ≈ 16.41ms, new ≈
16.98ms — the new version is actually slightly *slower*, not faster,
though the gap is small enough it may just be noise.

Dropped the runtime claim from the description and corrected it with
the real numbers. The verified win here is memory: 122.07 → 68.79 MiB,
43.6% less, which reproduces consistently across several scales.

Also pushed:
- The end-to-end dtype test (nullable Int64, string, categorical,
  datetime) through `pivot_longer(..., sort_by_appearance=True)`,
  covering the "tests added" claim from earlier
- A `reps=0` edge case test for defensive coverage
- Lint/formatting fixes (import ordering, line length)

Updated the PR description to match. Let me know if you'd like
anything else adjusted.

**Comment by samukweku** [comment]
@Anupam2400 I stress-tested this with very tall, very wide, and empty shapes. The new approach is a good win for normal rectangles, but it needs fast paths before merging.

Think of `rows` and `columns` as two rulers used to build the answer. For a normal rectangle, those small rulers are cheaper than keeping two full answers. But when there is only one row or one column, one ruler is as large as the answer, so the PR uses about twice the memory:

| Shape | Current | PR |
|---|---:|---:|
| `1 x 50,000,000` | 381.5 MiB | 762.9 MiB |
| `50,000,000 x 1` | 381.5 MiB | 762.9 MiB |
| `0 x 50,000,000` | ~0 MiB | 381.5 MiB |

For comparison, the intended case works well: `10,000,000 x 10` fell from 1,525.9 MiB to 839.4 MiB.

Suggested shape of the fix:

```python
def _build_indexer_reorder_contents(length: int, reps: int) -> np.ndarray:
    # Keep/validate the existing error contract separately for reps <= 0.
    if length <= 1 or reps == 1:
        return np.arange(length * reps)

    rows = np.arange(length)[:, None]
    columns = np.arange(reps) * length
    return np.add(rows, columns).ravel()
```

Please also add memory tests for `0 x N`, `1 x N`, and `N x 1`. The current `2 x 1000` / `1000 x 2` tests miss this boundary.

One separate behavior issue: the old helper raises `ValueError` for `reps=0` or negative `reps`, while this PR silently returns an empty array. Issue #1655 asks to preserve error behavior, and this is a performance-only PR, so please preserve the exception or add explicit validation instead of changing the contract here.

Correctness otherwise looked good: 920 valid shapes matched element-for-element, the 101 focused pivot tests passed, and large end-to-end nullable-`Int64` pivots preserved values and dtype.

**Review by samukweku — CHANGES_REQUESTED** [review]
left some comments @Anupam2400

**Comment by Anupam2400** [comment]
Thanks @samukweku  for the stress-testing — makes sense once traced through: at `length<=1` 
or `reps==1`, one of `rows`/`columns` is already the full output size, so 
broadcasting materializes a second copy.

**Pushed the following fixes:**

1. **Fast path** for `length <= 1` or `reps == 1` — returns `np.arange(length * reps)` 
   directly, avoiding the broadcast double-allocation

2. **Restored `reps <= 0` → `ValueError`** — removed the test that had locked in 
   the silent-empty regression, added explicit tests for zero and negative reps

3. **Parametrized memory tests** at the `1×N`, `N×1`, `0×N` boundaries

**Reproduced your exact numbers locally:**

| Shape | Old PR | After fix |
|---|---|---|
| `1 × 50,000,000` | 762.9 MiB | 381.5 MiB ✅ |
| `50,000,000 × 1` | 762.9 MiB | 381.5 MiB ✅ |
| `0 × 50,000,000` | 381.5 MiB | ~0 MiB ✅ |

105 `test_pivot_longer` tests pass locally.

**Comment by samukweku** [comment]
@Anupam2400 the boundary fixes work, and I found no remaining value/order/dtype mismatch. One performance blocker remains on the current head (`6498479`).

ELI5: the new indexer is like making a row ruler plus a column ruler. That saves memory when there are many columns/repetitions, but on a very tall frame with only 2–4 repetitions, building and broadcasting the tall ruler costs more time. The finished `pivot_longer` output already dominates memory, so that helper saving disappears end to end.

Current-head measurements (median of repeated runs, same pandas/pixi environment):

| Case | `dev` | PR | Result |
| --- | ---: | ---: | --- |
| helper, `1,000,000 x 2` | 2.96 ms | 5.18 ms | 75% slower |
| helper, `1,000,000 x 3` | 3.66 ms | 6.25 ms | 71% slower |
| helper, `1,000,000 x 4` | 5.88 ms | 7.03 ms | 20% slower |
| nullable-`Int64` `pivot_longer`, 1M rows x 2 reps | 14.05 ms | 16.85 ms | 20% slower |
| string `pivot_longer`, 1M rows x 2 reps | 14.28 ms | 16.47 ms | 15% slower |

For those 1M-row public calls, peak Python allocation was effectively identical (`49.61 MiB` for nullable integers and `30.53 MiB` for strings). At 5M nullable rows the PR was still 11% slower (`77.09 -> 85.39 ms`) with the same ~`248 MiB` peak.

So the helper-only memory test at `reps=8` does not cover issue #1655's end-to-end requirement or the shape where the algorithm loses. The `reps=8` case is genuinely good; the problem is choosing one formula for every non-degenerate shape.

Smallest likely fix: retain the old reshape/Fortran-ravel path for tall inputs with few repetitions, and use the broadcast path once repetitions are large enough to pay for it. Please choose the cutoff from a small shape matrix rather than hard-coding from one machine, and add an end-to-end tall/2-repetition benchmark alongside the helper memory benchmark. A CI timing assertion is unnecessary; recorded comparative benchmark evidence is enough.

**Comment by samukweku** [comment]
@Anupam2400 I left some comments; I'd suggest you run an adversarial review of your PR (and possibly future PRs) to get more feedback (play with different shapes and sizes, find best and worst cases, edge cases, ...)

**Comment by Anupam2400** [comment]
@samukweku — ran the adversarial benchmark across lengths and reps.

The crossover is not clean — it varies by both `length` and `reps`, 
and likely by CPU/cache as well:

| Shape | Old | New | Winner |
|---|---|---|---|
| 10000×2 | 0.091ms | 0.070ms | NEW |
| 5000×2 | 0.025ms | 0.035ms | OLD |
| 1000000×3 | 11.1ms | 7.6ms | NEW |
| 500000×3 | 2.58ms | 3.67ms | OLD |
| 5000000×2 | 41.6ms | 46.1ms | OLD |

The `reps=2-4, length<1M` range is where OLD consistently wins on 
my machine. For `length>=1M` with `reps>=2`, NEW wins.

Would a threshold like `if length < 1_000_000: use old path` make 
sense here? Or would you prefer a different approach — e.g. based on 
total output size `length * reps`?

Happy to implement whichever cutoff you think is most robust.

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,6 +1,7 @@
 # Changelog
 
 ## [Unreleased]
+-   [PERF] Reduce peak memory in `_build_indexer_reorder_contents` by ~44% using single NumPy allocation. - Issue #1655 @Anupam2400
```
Comment (samukweku) [inline CHANGELOG.md]: peak memory can vary based on frame's shape/size/counts etc; let's reframe the changelog entry here @Anupam2400

**Comment by samukweku** [comment]
@Anupam2400 let's use reps as the cutoff instead

gate on reps >= 8 — old path below that, new path at/above it. That's squarely in the "unambiguously good" zone across both benchmark sets, sidesteps the noisy length-dependent crossover entirely, and satisfies the issue as written without more benchmark iteration.

**Comment by Anupam2400** [comment]
@samukweku — implemented reps>=8 cutoff as suggested.

- Correctness verified at boundary (reps=6,7,8,9 all match element-for-element)
- Added two boundary tests: reps=7 (old path) and reps=8 (new path)
- Updated CHANGELOG to remove shape-dependent percentage claim
- 107 test_pivot_longer tests pass

File: janitor/functions/pivot.py [hunk 2]
```diff
@@ -1831,11 +1831,39 @@ def _build_indexer_for_spec(
...
+    twice the output size.
+
+    For tall frames with few repetitions (reps < 8), retains the
+    original reshape/Fortran-ravel path which is faster in this
+    regime since building and broadcasting the tall rows ruler
+    costs more than the temporary copy.
+    """
+    if reps <= 0:
+        raise ValueError(f"reps must be a positive integer, got {reps}.")
+    if length <= 1 or reps == 1:
```
Comment (samukweku) [inline janitor/functions/pivot.py]: it isnt possible for length to be less than 1.

Comment (samukweku) [reply janitor/functions/pivot.py]: thinking about it, i'm not sure reps can be less than 0. at any rate, i'd suggest you move the length check to the `reps <= 0` entry above (`reps <=0 or length <1` or something similar)

**Review by samukweku — CHANGES_REQUESTED** [review]

File: janitor/functions/pivot.py (hunk 2 above)
Comment (Anupam2400) [reply janitor/functions/pivot.py]: Thanks for the suggestion! Moved the length check up to sit right next to the reps <= 0 guard. Since length < 1 is a valid degenerate case (empty frame → return empty array) while reps <= 0 is an invalid input (→ ValueError), they need separate branches, so the result looks like:

if reps <= 0:
    raise ValueError(f"reps must be a positive integer, got {reps}.")
if length < 1 or reps == 1:
    return np.arange(length * reps)


Also changed length <= 1 to length < 1 since length == 1 is handled correctly by both downstream paths without the shortcut. All 107 tests pass.

File: janitor/functions/pivot.py [hunk 3]
```diff
@@ -1853,7 +1853,7 @@ def _build_indexer_reorder_contents(length: int, reps: int) -> np.ndarray:
     """
     if reps <= 0:
         raise ValueError(f"reps must be a positive integer, got {reps}.")
-    if length <= 1 or reps == 1:
+    if length < 1 or reps == 1:
```
Comment (samukweku) [inline janitor/functions/pivot.py:1856]: ~if length is less than 1, then np.arange should return nothing. doesnt sound right.~

**Review by samukweku — APPROVED** [review]

**Comment by samukweku** [comment]
thanks @Anupam2400
